pkg/apis/networking/types.go

```diff
@@ -246,16 +246,39 @@ type IngressList struct {
 
 // IngressSpec describes the Ingress the user wishes to exist.
 type IngressSpec struct {
-	// IngressClassName is the name of the IngressClass cluster resource. The
-	// associated IngressClass defines which controller will implement the
-	// resource. This replaces the deprecated `kubernetes.io/ingress.class`
-	// annotation. For backwards compatibility, when that annotation is set, it
-	// must be given precedence over this field. The controller may emit a
-	// warning if the field and annotation have different values.
-	// Implementations of this API should ignore Ingresses without a class
-	// specified. An IngressClass resource may be marked as default, which can
-	// be used to set a default value for this field. For more information,
-	// refer to the IngressClass documentation.
+	// IngressClassName is the name of an IngressClass cluster resource.
+	// Ingress controller implementations use this field to know whether they
+	// should be picking up this Ingress.
+	//
+	// The field `ingressClassName` (referred to as “the field”) should be
+	// preferred over the annotation `kubernetes.io/ingress.class` (referred to
+	// as “the annotation”). The annotation and the field are different: the
+	// field refers to an IngressClass that exists in the cluster (although
+	// implementations may not require the presence of an IngressClass with a
+	// matching name). On the other side, the annotation is an opaque string
+	// that does not relate to the IngressClass type.
+	//
+	// The annotation and the field cannot be set simultaneously when an Ingress
+	// is created. In order to ease the process of progressively migrating from
+	// the annotation to the field, it is still possible to set both the
+	// annotation and field by updating an existing Ingress. This is useful for
+	// people migrating away from an implementation that does not support the
+	// field. Because of this possibility of having both the annotation and
+	// field, implementations must give the annotation precedence over the field
+	// when both are set. Kubernetes does not validate that both the annotation
+	// and field match when an Ingress is updated; having mismatched values may
+	// be useful for migrating. Implementations may warn the user of such a
+	// mismatch.
+	//
+	// An IngressClass resource may be marked as default by setting the
+	// annotation `ingressclass.kubernetes.io/is-default-class=true` on the
+	// IngressClass. When this annotation is set on an IngressClass, Kubernetes
+	// will look for Ingresses that are being created without the field, and
+	// default the field to the name of that IngressClass. If multiple
+	// IngressClasses are set as default, Kubernetes stops defaulting Ingresses.
+	// For backwards compatibility with implementations that do not support the
+	// field yet, Kubernetes won’t default Ingresses that already have the
+	// annotation set.
```

Comment:
Instead of this text, how about recommending that readers refer to the IngressClass API for relevant details?

API reference documentation is not expected to provide broader context (for that, we have the [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) concept page).

Reply:
It is OK to put this text in the API details for IngressClass.